torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+            raise ValueError(f"unsupported dtype (SympyToZ3): {dtype}")
+
+        def truediv(self, numerator: z3.ArithRef, denominator: z3.ArithRef) -> z3.ArithRef:
+            return self._add_assertion_and_arith(Z3Ops.div(numerator, denominator))
+
+        def div(self, numerator: z3.ArithRef, denominator: z3.ArithRef) -> z3.ArithRef:
+            return self._add_assertion_and_arith(Z3Ops.floordiv(numerator, denominator))
+
+        def pow(self, base: z3.ArithRef, exp: z3.ArithRef) -> z3.ArithRef:
+            return self._add_assertion_and_arith(Z3Ops.pow(base, exp))
+
+        def mod(self, p: z3.ArithRef, q: z3.ArithRef) -> z3.ArithRef:
+            return self._add_assertion_and_arith(Z3Ops.mod(p, q))
+
+        def __getattr__(self, name: str) -> Any:
+            if name in self.Z3_REPLACEMENT_HANDLES:
+                return self.Z3_REPLACEMENT_HANDLES[name]
+            if name in self.OPERATOR_HANDLES:
+                return getattr(operator, name)
+            print(name)
+            raise AttributeError
+
+        def run(self, expr: sympy.Basic) -> Z3Expr:
+            return sympy_interp(self, self._validator.symbols, expr)  # type: ignore[arg-type]
+
+    # Frontend class to Z3 validation.
+    #
+    # Given:
+    # - Input: all the generated SymPy expressions (no optimizations applied)
+    # - Output: optimized SymPy expressions
+    #
+    # Shows whether the transformations applied to the input expressions are
+    # sound. In other words, solves the following soundness problem:
+    #
+    # (for all possible assignments of the free variables)
+    #
+    # "If the conjunction of the output expressions is true, is the conjunctions
+    # of the input expressions also true?"
+    #
+    # In practice: if there is any assignment of the free variables such that
+    # the conjunction of the inputs is FALSE but the conjunction of the outputs
+    # is TRUE, it would mean that an unsound transformation was applied at some
+    # point.
+    #
+    # That is exactly what this class does. Checks whether the following holds
+    # for any assignment of free variables:
+    #
+    # Not(And(Input)) AND And(Output)
+    #
+    # If the above equation is true, it means that there is a case where the
+    # generated guard (using the output expressions) is incorrectly true.
+    # Otherwise, it shows that the optimized output expressions is sound.
+    class TranslationValidator:
+        def __init__(self) -> None:
+            super().__init__()
+
+            # Mapping of SymPy symbols to Z3 integer variables.
+            self.symbols = {}
+
+            # Set of Z3 expressions representing all the generated guards.
+            # i.e. without any transformation/skipping.
+            self._inputs = set()
+
+            # Set of Z3 expressions representing the actual issued guards.
+            # i.e. those that will actually be used in the guard function.
+            self._outputs = set()
```

Comment:
It might be better to stop using the input/output naming and do something like original/optimized

Reply:
LLVM Alive2 uses "source" and "target" nomenclature.